cmd/kubeadm/app/util/staticpod/utils.go

```diff
@@ -97,6 +97,24 @@ func ComponentProbe(cfg *kubeadmapi.MasterConfiguration, componentName string, p
 	}
 }
 
+// EtcdProbe is a helper function for building a shell-based, etcdctl v1.Probe object to healthcheck etcd
+func EtcdProbe(cfg *kubeadmapi.MasterConfiguration, componentName string, port int, certsDir string, CACertName string, CertName string, KeyName string) *v1.Probe {
+	tlsFlags := fmt.Sprintf("--cacert=%[1]s/%[2]s --cert=%[1]s/%[3]s --key=%[1]s/%[4]s", certsDir, CACertName, CertName, KeyName)
+	// etcd pod is alive if a linearizable get succeeds.
+	cmd := fmt.Sprintf("ETCDCTL_API=3 etcdctl --endpoints=%s:%d %s get foo", GetProbeAddress(cfg, componentName), port, tlsFlags)
```

Comment:
This pre-supposes that etcdctl is installed and it's not part of the installation process.  You'd have to use the client library.

Reply:
Hi Tim,
Would you please clarify? The liveness probe exec's into the container:
https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.9/#execaction-v1-core

This does assume that `etcdctl` is an included binary on the `PATH` inside the `etcd` container, and that it's capable of using the V3 API.

Reply:
I have verified this works with the current upstream image: `k8s.gcr.io/etcd-amd64:3.1.11`